### ISSUE TITLE
How pass array to oracle procedure using in-parameter?

### ISSUE THREAD

**Issue by alciakng** [oracle/node-oracledb#98]
I try passing array to oracle procedure in node.js.
below is my code. but failure. help me.. 

```javascript
var BIGINBLOCK ="declare "
       +"p_seatcodes dbms_sql.varchar2a; "
       +"begin "
       +"p_seatcodes := :1; " 
       +"RESERVEPROC(:p_email,:p_class,:p_bookingcode,:p_timecode,:p_screencode,:p_moviecode,:p_totalprice,:p_seatcount,p_seatcodes); "
       +"end;"
```

and my procedure is 
```sql
CREATE OR REPLACE PROCEDURE RESERVEPROC
( p_email in varchar2,
  p_class in varchar2,
  p_bookingcode in varchar2,
  p_timecode in varchar2,
  p_screencode in varchar2,
  p_moviecode in varchar2,
  p_totalprice in integer,
  p_seatcount in integer,
  p_seatcodes in dbms_sql.varchar2a
 )
 IS 
    I INTEGER;
 BEGIN
    --멤버 마일리지 업데이트
    UPDATE MEMBER
    SET MILEAGE = CASE 
                   WHEN class = 'D' THEN MILEAGE+p_totalprice*0.02
                   WHEN class = 'C' THEN MILEAGE+p_totalprice*0.04
                   WHEN class = 'B' THEN  MILEAGE+p_totalprice*0.06
                   WHEN class = 'A' THEN  MILEAGE+p_totalprice*0.08
                   WHEN class = 'S' THEN  MILEAGE+p_totalprice*0.1
                   END
    WHERE EMAIL = p_email;
                   
    --BOOKING테이블 예약레코드 삽입               
    INSERT
    INTO BOOKING 
    VALUES (p_bookingcode,p_email,p_timecode,p_screencode,p_moviecode,p_totalprice,p_seatcount);               
     
    --seatcodearr을 루프 돌면서 BOOKED_SEATS에 삽입하고 PERFORMANCE_SEAT의 사용여부 업데이트.  
    FOR I IN 1 .. p_seatcodes.COUNT
    LOOP
        --BOOKED_SEATS에 INSERT
        INSERT
        INTO BOOKED_SEATS 
        VALUES(p_bookingcode,p_seatcodes(I));
         
        --PERFORMANCE_SEAT UPDATE
        UPDATE
        PERFORMANCE_SEAT 
        SET SEATSTATUS=1 
        WHERE SEATCODE= p_seatcodes(I)
        and TIMECODE= p_timecode;
    END LOOP;
    --커밋.
    commit;
END;
```

**Comment by cjbj** [comment]
Support for binding PL/SQL arrays has not yet been implemented.

**Comment by krish0803** [comment]
Is this feature implemented in v0.7?

I need to pass an array of strings as input parameter. 
Below is the sample of ARRAY defined in DB.

CREATE OR REPLACE
TYPE VAR_ARRAY AS TABLE OF VARCHAR2(100);

**Comment by cjbj** [comment]
@krish0803 The list of changes is in https://github.com/oracle/node-oracledb/blob/master/CHANGELOG.md

Your request is not on the roadmap for <= 1.0.  We'll evaluate priorities again after that release.
[The things I personally want to have in by 1.0 are LOBs, some basic type mapping for queries, a review of date handling, and various NLS issues fixed]

**Comment by lewisbarber** [comment]
This would be a great feature to get in there. From an app development perspective, it significantly enables me to improve performance, if I can send 10k entries to the database for insert / update in one go. Considering porting a Java app to Node (with Oracle DB), and this missing functionality might be enough to stop me from doing so until it's implemented. If only I knew C++ well enough, I'd try and implement it myself :+1:

**Comment by SirRagnar** [comment]
+1
This feature would be awesome.

**Comment by doberkofler** [comment]
This would be the most important missing feature for me and in fact prevents me to switch from my own nodejs driver to the "official" one.

**Comment by lewisbarber** [comment]
@doberkofler do you have a Nodejs driver that supports this functionality? If you do I'd love it if you could share it with me whilst we're waiting for this functionality to be implemented. Thanks

**Comment by doberkofler** [comment]
@lewisbarber Yes, I do actually have a driver that supports this functionality, but it is not a generic one but rather specialized for my https://github.com/doberkofler/node_plsql project.
I'm currently evaluating switching to to oracledb and would most likely then open a pull request with this functionality.

**Comment by cjbj** [comment]
@doberkofler we'd be glad to see a PR.  Before you start work check the [CONTRIBUTING file](https://github.com/oracle/node-oracledb/blob/master/CONTRIBUTING.md) and post a draft plan so we can see where it is headed.

**Comment by hg-smart** [comment]
Thats exactly what I need for performing a bulk insert into the database, passing an array to a stored procedure in the database from node.js and then performing a bulk insert using a forall insert construct...

**Comment by doberkofler** [comment]
@cjbj I started analyzing the existing code base and am currently preparing a prototype for passing arrays to stored procedures. Because I'm using node 4, all my work is currently based on the nan 2 pull request #227, so I will not be able to create a pull request before #227 has been merged.
My current plan is to enhance the object bindings and simply allow to set the 'val' property to an array of values instead of just a plain value.

An example for the binding would look as follows:
```
var bindvars = {
  'strings': {type: oracledb.STRING, dir: oracledb.BIND_IN, maxSize: 30, val: ['John', 'Doe']},
  'numbers': {type: oracledb.NUMBER, dir: oracledb.BIND_IN, val: [1, 2, 3, 4, 5]}
};
connection.execute('BEGIN testproc(:strings, :numbers); END;', bindvars, function (err, result) {
  if (err) {
    console.error(err.message);
    return;
  }
  console.log(result.outBinds);        
});
```

**Comment by lewisbarber** [comment]
@doberkofler - this would be great but it would also be useful to be able to pass an ARRAY of OBJECTS to the server, for instance, using the code sample above I was thinking:

var bindvars = {
  'objects': {type: oracledb.OBJECT, dir: oracledb.BIND_IN, maxSize: 30, val: [{name:'John', age:20}, {name:'Doe',age:50}]}
};
connection.execute('BEGIN testproc(:objects); END;', bindvars, function (err, result) {
  if (err) {
    console.error(err.message);
    return;
  }
  console.log(result.outBinds);        
});

@cjbj - should we open another issue for support for passing objects (custom types) to stored procs or can this issue be used for this as well?

Thanks

**Comment by doberkofler** [comment]
@lewisbarber I'm currently not planning to support this but maybe others will. I'm actually wondering, if more complex data structures should not rather be encoded as JSON strings altogether but this is rather a strategic topic that @cjbj might be able to answer. I did start by supporting binding arrays of values for `IN` parameters and in a second step would also like to support `out` and `IN OUT` bindings.

**Comment by cjbj** [comment]
@doberkofler OUT binds should be part of the initial API design, even if implementation of that bit is deferred.  You are using `maxSize` which is currently for OUT binds.  Did you mean to have it in the example?  What types are you proposing to support?

I'd like to see other people's comments on your design.  

Can you make sure you start the [OCA](https://github.com/oracle/node-oracledb/blob/master/CONTRIBUTING.md) paperwork?

Good work.

@lewisbarber I was using #147 for custom type support. We're open to ideas about the design.

**Comment by doberkofler** [comment]
@cjbj to your comments:
* I have just filed the OCA paperwork
* support for OUT binds is planned but the implementation deferred
* I used maxSize on purpose and out of principle because array bind do actually always need a fixed length but now that you asked, I also see the advantage in just calculating maxSize dynamically when dealing with an IN bind.
* I would want to support the 3 plain JavaScript types (string, number, boolean) and for convenience also explicitly the Date object.

**Comment by doberkofler** [comment]
@cjbj To offer a more comprehensive example of what my current implementation would look like, I can show how the unit test looks like:
```javascript
    it('4.6.1 binding arrays of values with different types', function(done) {
      async.series([
        function(callback) {
          var proc = "CREATE OR REPLACE PACKAGE \
                      oracledb_testpack \
                      IS \
                        TYPE stringTable IS TABLE OF VARCHAR2(30) INDEX BY BINARY_INTEGER; \
                        TYPE numberTable IS TABLE OF NUMBER INDEX BY BINARY_INTEGER; \
                        FUNCTION test(strings IN stringTable, numbers IN numberTable) RETURN VARCHAR2; \
                      END;";
          connection.should.be.ok;
          connection.execute(
            proc,
            function(err) {
              should.not.exist(err);
              callback();
            }
          );
        },
        function(callback) {
          var proc = "CREATE OR REPLACE PACKAGE BODY \
                      oracledb_testpack \
                      IS \
                        FUNCTION test(strings IN stringTable, numbers IN numberTable) RETURN VARCHAR2 \
                        IS \
                          s VARCHAR2(2000) := ''; \
                        BEGIN \
                          FOR i IN 1 .. strings.COUNT LOOP \
                            s := s || strings(i); \
                          END LOOP; \
                          FOR i IN 1 .. numbers.COUNT LOOP \
                            s := s || numbers(i); \
                          END LOOP; \
                          RETURN s; \
                        END; \
                      END;";
          connection.should.be.ok;
          connection.execute(
            proc,
            function(err) {
              should.not.exist(err);
              callback();
            }
          );
        },
        function(callback) {
          var bindvars = {
            result: {type: oracledb.STRING, dir: oracledb.BIND_OUT, maxSize: 2000},
            strings:  {type: oracledb.STRING, dir: oracledb.BIND_IN, val: ['John', 'Doe']},
            numbers: {type: oracledb.NUMBER, dir: oracledb.BIND_IN, val: [0, 8, 11]}
          };
          connection.execute(
            "BEGIN :result := oracledb_testpack.test(:strings, :numbers); END;",
            bindvars,
            function(err, result) {
              should.not.exist(err);
              // console.log(result);
              result.outBinds.result.should.be.exactly('JohnDoe0811');
              callback();
            }
          );
        }
      ], done);
    })
```

**Comment by cjbj** [comment]
@doberkofler we briefly reviewed the API today, thinking about how IN OUT and OUT PL/SQL binds would work.  We also wanted to make sure that any future Array DML bind support wouldn't cause the API to become inconsistent.  At first glance, all looks good.  For IN OUT & OUT PL/SQL binds we could augment your IN design with a new, bind `maxArraySize` attribute to say how many elements should be expected on return.  We may uncover things when we get a chance to look at the implementation, but for now, I'd proceed.  I heard you need to resubmit your OCA application?

**Comment by cjbj** [comment]
@doberkofler you mentioned boolean support.  Do you need this or is it a 'nice to have'?

**Comment by doberkofler** [comment]
@cjbj 
* Yes, for OUT and IN/OUT bind we definitively need a new property for maximal array size and `maxArraySize` sound perfect to me
* I wanted to finalize and test my implementation for IN bind before looking at anything else and as far as I can currently tell, it should be ok and not be intrusive at all.
* Just forget boolean support. (I overlooked that non-array binds also do not support it)
* Yes, I had to resubmit the OCA application because I typically do not give away my phone number and mail address but I made an exception and hope for the best ;-)

**Comment by cjbj** [comment]
@doberkofler thanks for the clear status.  We don't share email addresses.

**Comment by cjbj** [comment]
@doberkofler how are you going with this?

**Comment by doberkofler** [comment]
@cjbj I'm currently on hold because all my work is currently based on the nan 2 pull request #227, so I will not be able to create a pull request before #227 has been merged.

**Comment by cjbj** [comment]
@doberkofler no problem.  Let's hope we can pin down the NAN 2 mem leak issue soon so #227 can get merged.

**Comment by doberkofler** [comment]
@cjbj I will now continue working on this

**Comment by cjbj** [comment]
@doberkofler excellent.

**Comment by doberkofler** [comment]
@cjbj I've now spend a lot of time in understanding how the binding of date/timestamp values is done in node-oracledb and came to the conclusion that the OCIDateTime types that are always used when binding a date or a timestamp do not seem to allow the binding of a PL/SQL indexed table. I'm currently waiting for an (overdue) feedback to a Oracle SR and wanted to know if you or anyone else has more information on this topic.

**Comment by doberkofler** [comment]
@cjbj I am still waiting for feedback on my SR, but for now, it looks as if biding an OCIDateTime as an PL/SQL indexed table seem not to be supported.
While playing around with all kinds of test apps to track down this problem, I realized that there are only a few possible workarounds and I would like to hear your opinion on them:
* Simply not allow at all date/timestamp values to be bound an a PL/SQL indexed table (not very nice but most likely not a serous restriction in real applications)
* Use a simple SQLT_DAT binding either only for PL/SQL indexed table binds or generally (maybe you count give me some background information for the need of using the SQLT_TIMESTAMP_LTZ data type)

**Comment by cjbj** [comment]
@doberkofler  I'm fine with looking at dates & ts's later.  PS I sent you email a few days back.

**Comment by doberkofler** [comment]
@cjbj I've received your email and did update my original post with the number of the SR

**Comment by cjbj** [comment]
@doberkofler I'l review it with the team

**Comment by doberkofler** [comment]
@cjbj Thank you and just to keep you posted: After I've stopped trying to bind indexed tables of date/timestamp values, I was able to finalize the binding of indexed tables for IN and IN OUT bind variables including a set of unit tests in no time. The final part I'm currently looking at are the OUT bind variables but as for now this seems to be a lot more intrusive and might take a little longer to get working as expected.

**Comment by cjbj** [comment]
@doberkofler excellent.  Is there some code etc you want us to look at before you get too far?

**Comment by doberkofler** [comment]
@cjbj My code is a little messy right now with a ton of debug and trace messages all over the place. Let me try to see if a can finish implementing OUT variables within the next week and then issue a pull request for all the changes. If it takes longer, I would push the the changes for IN and IN OUT separately.

**Comment by doberkofler** [comment]
@cjbj I was now able to finalize my work and my commit https://github.com/doberkofler/node-oracledb/commit/625538a5c4b8b8fca4e823bb2dd83a0670120da1 contains all the needed changes to support binding pf PL/SQL indexes table values. I already did also add a unit test but not change the documentation.

**Comment by cjbj** [comment]
@doberkofler that's a lot of good progress.  I don't mind helping with the doc when the time comes.  What would be useful when you are ready is a brief summary of what is new, of what is & isn't expected to work, and about anything that could do with a deeper review.  PS can you use NJS-033, since 30-32 will be used in 1.5 for messages about preventing connection release while DB operations are in progress?

**Comment by doberkofler** [comment]
@cjbj

Summary:
* the current implementation support binding of PL/SQL indexed tables
* arrays containing strings and numbers are supported
* because of an OCI problem dates are not yet supported
* only the object syntax of bindings is allowed 
* IN, IN OUT and OUT binding are fully supported
* IN bindings only need a val property defined as an array to be identified as array binding
* IN OUT and OUT binding additionally need the maxSize and maxArraySize properties to be specified
* The new unit test PLSQLbinds.js shows all types of binding and also tests for all possible exceptions
* The documentation for the new functionality has not been finalized

Error message:
Regarding the error messages, I see no proper way on how to "reserve" some messages with the current "flat" data structure and the only thing, I could do is to change the actual error text from '"NJS-030: invalid (array) binding of parameter \"%s\": %s"' to '"NJS-033: invalid (array) binding of parameter \"%s\": %s"', is this is what you mean.
I generally would have preferred to add my new error message where the other bind messages are and have them all grouped together bud did not want to change the existing data structure without having this discussed first.

**Comment by cjbj** [comment]
@doberkofler Nice.  And don't worry about error numbers now (though we are organically heading towards the same situation as https://github.com/oracle/node-oracledb/issues/308 - suggestions welcome)  Are you going to submit a PR?

**Comment by doberkofler** [comment]
@cjbj Done! I just wanted to do some more testing on a larger scale before submitting my PR.

**Comment by cjbj** [comment]
@doberkofler thanks very much!  We'll look at this for 1.6 (because 1.5 is turning into a bug fix release and I don't want to delay it)

Can you add the signoff as mentioned in https://github.com/oracle/node-oracledb/blob/master/CONTRIBUTING.md ?

**Comment by doberkofler** [comment]
@cjbj Just added the missing signoff

**Comment by cjbj** [comment]
Thank you

**Comment by cjbj** [comment]
Closing this issue which devolved mostly into a discussion of Index-by support subsequently implemented in node-oracledb 1.7.1 thanks to @doberkofler.  The related issue #348 is still open.

**Comment by cjbj** [comment]
For users trying to call one PL/SQL block repeated times with different parameters, check out node-oracledb 2.2's [executeMany()](https://oracle.github.io/node-oracledb/doc/api.html#executemany) call.  There is a PL/SQL example [here](https://github.com/oracle/node-oracledb/blob/master/examples/em_plsql.js).
